pkg/util/openshift.go

```diff
@@ -1,68 +1,90 @@
 /*
- * Copyright 2019, EnMasse authors.
+ * Copyright 2019-2020, EnMasse authors.
  * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
  */
 package util
 
 import (
+	"context"
 	"encoding/json"
 	"fmt"
 	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"net/url"
 	"os"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"strings"
 	"time"
 
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 
+	routeapiv1 "github.com/openshift/api/route/v1"
 	routev1 "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 )
 
 var (
 	openshift *bool
+	apis      = make(map[string]bool, 0)
 	log       = logf.Log.WithName("util")
 )
 
 const ConnectsTo = "app.openshift.io/connects-to"
 
 func IsOpenshift() bool {
+
 	if openshift == nil {
 		b := detectOpenshift()
 		openshift = &b
 	}
+
 	return *openshift
+
 }
 
-func detectOpenshift() bool {
+func HasApi(gvk schema.GroupVersionKind) bool {
+	if has, ok := apis[gvk.String()]; ok {
+		return has
+	}
 
-	log.Info("Detect if openshift is running")
+	has := detectApi(gvk)
+	apis[gvk.String()] = has
+	return has
+}
 
-	value, ok := os.LookupEnv("ENMASSE_OPENSHIFT")
+func detectApi(gvk schema.GroupVersionKind) bool {
+
+	name := fmt.Sprintf("ENMASSE_HAS_API_%s_%s_%s", strings.ToUpper(gvk.Group), strings.ToUpper(gvk.Version), strings.ToUpper(gvk.Kind))
+
+	value, ok := os.LookupEnv(name)
 	if ok {
-		log.Info("Set by env-var 'ENMASSE_OPENSHIFT': " + value)
+		log.Info("Set by env-var '" + name + "': " + value)
 		return strings.ToLower(value) == "true"
```

Comment:
Should we be using `strconv.ParseBool()` (and treating error as false)?  A user familiar to Go conventions could pass one of the other Go truthy values (1, T, 0, F)?

Reply:
I kept that line as it was, but I agree, that might be a nice change.

Reply:
Just checked on the API, that would require us to handle `error` was well … which is different than the Java side